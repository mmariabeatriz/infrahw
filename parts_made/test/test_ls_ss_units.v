// Teste das unidades LS e SS adaptadas conforme especificação
// Este teste verifica se as unidades LS e SS estão funcionando corretamente
// com os novos sinais de controle de 2 bits

`timescale 1ns / 1ps

module test_ls_ss_units;

    // Sinais para teste da unidade LS
    reg [1:0] ls_control;
    reg [31:0] reg_mdr_out;
    wire [31:0] ls_control_out;
    
    // Sinais para teste da unidade SS
    reg [1:0] ss_control;
    reg [31:0] reg_b_out;
    reg [31:0] reg_mdr_out_ss;
    wire [31:0] ss_control_out;
    
    // Instanciação da unidade LS
    ls uut_ls (
        .LSControl(ls_control),
        .RegMDROut(reg_mdr_out),
        .LSControlOut(ls_control_out)
    );
    
    // Instanciação da unidade SS
    ss uut_ss (
        .RegSSControl(ss_control),
        .RegBOut(reg_b_out),
        .RegMDROut(reg_mdr_out_ss),
        .SSControlOut(ss_control_out)
    );
    
    // Configuração para dump VCD
    initial begin
        $dumpfile("test_ls_ss_units.vcd");
        $dumpvars(0, test_ls_ss_units);
    end
    
    // Testes principais
    initial begin
        $display("=== TESTE DAS UNIDADES LS E SS ADAPTADAS ===");
        
        // === TESTE DA UNIDADE LS ===
        $display("\n=== TESTE DA UNIDADE LS (Load Size) ===");
        
        // Valor de teste na memória
        reg_mdr_out = 32'hAABBCCDD;
        
        // Teste Load Byte (LSControl = 00)
        ls_control = 2'b00;
        #10;
        $display("Load Byte (LSControl=00):");
        $display("  RegMDROut: 0x%h", reg_mdr_out);
        $display("  LSControlOut: 0x%h (esperado: 0x000000DD)", ls_control_out);
        
        // Teste Load Halfword (LSControl = 01)
        ls_control = 2'b01;
        #10;
        $display("Load Halfword (LSControl=01):");
        $display("  RegMDROut: 0x%h", reg_mdr_out);
        $display("  LSControlOut: 0x%h (esperado: 0x0000CCDD)", ls_control_out);
        
        // Teste Load Word (LSControl = 10)
        ls_control = 2'b10;
        #10;
        $display("Load Word (LSControl=10):");
        $display("  RegMDROut: 0x%h", reg_mdr_out);
        $display("  LSControlOut: 0x%h (esperado: 0xAABBCCDD)", ls_control_out);
        
        // === TESTE DA UNIDADE SS ===
        $display("\n=== TESTE DA UNIDADE SS (Store Size) ===");
        
        // Valores de teste
        reg_b_out = 32'h12345678;      // Novo valor a ser salvo
        reg_mdr_out_ss = 32'hAABBCCDD; // Valor antigo na memória
        
        // Teste Store Byte (SSControl = 00)
        ss_control = 2'b00;
        #10;
        $display("Store Byte (SSControl=00):");
        $display("  RegBOut: 0x%h", reg_b_out);
        $display("  RegMDROut: 0x%h", reg_mdr_out_ss);
        $display("  SSControlOut: 0x%h (esperado: 0xAABBCC78)", ss_control_out);
        
        // Teste Store Halfword (SSControl = 01)
        ss_control = 2'b01;
        #10;
        $display("Store Halfword (SSControl=01):");
        $display("  RegBOut: 0x%h", reg_b_out);
        $display("  RegMDROut: 0x%h", reg_mdr_out_ss);
        $display("  SSControlOut: 0x%h (esperado: 0xAABB5678)", ss_control_out);
        
        // Teste Store Word (SSControl = 10)
        ss_control = 2'b10;
        #10;
        $display("Store Word (SSControl=10):");
        $display("  RegBOut: 0x%h", reg_b_out);
        $display("  RegMDROut: 0x%h", reg_mdr_out_ss);
        $display("  SSControlOut: 0x%h (esperado: 0x12345678)", ss_control_out);
        
        // === TESTE DE CASOS EXTREMOS ===
        $display("\n=== TESTE DE CASOS EXTREMOS ===");
        
        // Teste com valores zero
        reg_mdr_out = 32'h00000000;
        reg_b_out = 32'h00000000;
        reg_mdr_out_ss = 32'hFFFFFFFF;
        
        ls_control = 2'b00; // Load byte de zero
        ss_control = 2'b00; // Store byte de zero
        #10;
        $display("Load Byte de zero: 0x%h (esperado: 0x00000000)", ls_control_out);
        $display("Store Byte de zero: 0x%h (esperado: 0xFFFFFF00)", ss_control_out);
        
        // Teste com valores máximos
        reg_mdr_out = 32'hFFFFFFFF;
        reg_b_out = 32'hFFFFFFFF;
        reg_mdr_out_ss = 32'h00000000;
        
        ls_control = 2'b10; // Load word máximo
        ss_control = 2'b10; // Store word máximo
        #10;
        $display("Load Word máximo: 0x%h (esperado: 0xFFFFFFFF)", ls_control_out);
        $display("Store Word máximo: 0x%h (esperado: 0xFFFFFFFF)", ss_control_out);
        
        $display("\n=== TESTE DAS UNIDADES LS E SS CONCLUÍDO ===");
        $finish;
    end
    
endmodule