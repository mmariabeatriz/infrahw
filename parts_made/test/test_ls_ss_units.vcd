$date
	Sat Jul 12 22:27:43 2025
$end
$version
	Icarus Verilog
$end
$timescale
	1ps
$end
$scope module test_ls_ss_units $end
$var wire 32 ! ss_control_out [31:0] $end
$var wire 32 " ls_control_out [31:0] $end
$var reg 2 # ls_control [1:0] $end
$var reg 32 $ reg_b_out [31:0] $end
$var reg 32 % reg_mdr_out [31:0] $end
$var reg 32 & reg_mdr_out_ss [31:0] $end
$var reg 2 ' ss_control [1:0] $end
$scope module uut_ls $end
$var wire 2 ( LSControl [1:0] $end
$var wire 32 ) RegMDROut [31:0] $end
$var reg 32 * LSControlOut [31:0] $end
$upscope $end
$scope module uut_ss $end
$var wire 32 + RegBOut [31:0] $end
$var wire 32 , RegMDROut [31:0] $end
$var wire 2 - RegSSControl [1:0] $end
$var reg 32 . SSControlOut [31:0] $end
$upscope $end
$upscope $end
$enddefinitions $end
$comment Show the parameter values. $end
$dumpall
$end
#0
$dumpvars
bx .
bx -
bx ,
bx +
b11011101 *
b10101010101110111100110011011101 )
b0 (
bx '
bx &
b10101010101110111100110011011101 %
bx $
b0 #
b11011101 "
bx !
$end
#10000
b1100110011011101 "
b1100110011011101 *
b1 #
b1 (
#20000
b10101010101110111100110011011101 "
b10101010101110111100110011011101 *
b10 #
b10 (
#30000
b10101010101110111100110001111000 !
b10101010101110111100110001111000 .
b0 '
b0 -
b10101010101110111100110011011101 &
b10101010101110111100110011011101 ,
b10010001101000101011001111000 $
b10010001101000101011001111000 +
#40000
b10101010101110110101011001111000 !
b10101010101110110101011001111000 .
b1 '
b1 -
#50000
b10010001101000101011001111000 !
b10010001101000101011001111000 .
b10 '
b10 -
#60000
b11111111111111111111111100000000 !
b11111111111111111111111100000000 .
b0 "
b0 *
b0 '
b0 -
b0 #
b0 (
b11111111111111111111111111111111 &
b11111111111111111111111111111111 ,
b0 $
b0 +
b0 %
b0 )
#70000
b11111111111111111111111111111111 !
b11111111111111111111111111111111 .
b11111111111111111111111111111111 "
b11111111111111111111111111111111 *
b10 '
b10 -
b10 #
b10 (
b0 &
b0 ,
b11111111111111111111111111111111 $
b11111111111111111111111111111111 +
b11111111111111111111111111111111 %
b11111111111111111111111111111111 )
#80000
