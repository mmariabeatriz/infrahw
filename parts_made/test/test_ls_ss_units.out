#! /c/Source/iverilog-install/bin/vvp
:ivl_version "12.0 (devel)" "(s20150603-1539-g2693dd32b)";
:ivl_delay_selection "TYPICAL";
:vpi_time_precision - 12;
:vpi_module "C:\iverilog\lib\ivl\system.vpi";
:vpi_module "C:\iverilog\lib\ivl\vhdl_sys.vpi";
:vpi_module "C:\iverilog\lib\ivl\vhdl_textio.vpi";
:vpi_module "C:\iverilog\lib\ivl\v2005_math.vpi";
:vpi_module "C:\iverilog\lib\ivl\va_math.vpi";
S_00000286ea6ef4b0 .scope module, "test_ls_ss_units" "test_ls_ss_units" 2 7;
 .timescale -9 -12;
v00000286ea6e6c00_0 .var "ls_control", 1 0;
v00000286ea6e6ca0_0 .net "ls_control_out", 31 0, v00000286ea6e8800_0;  1 drivers
v00000286ea6ec5a0_0 .var "reg_b_out", 31 0;
v00000286ea6ec640_0 .var "reg_mdr_out", 31 0;
v00000286ea6ec6e0_0 .var "reg_mdr_out_ss", 31 0;
v00000286ea6ec780_0 .var "ss_control", 1 0;
v00000286ea6ec820_0 .net "ss_control_out", 31 0, v00000286ea6e6b60_0;  1 drivers
S_00000286ea6abda0 .scope module, "uut_ls" "ls" 2 21, 3 14 0, S_00000286ea6ef4b0;
 .timescale -9 -12;
    .port_info 0 /INPUT 2 "LSControl";
    .port_info 1 /INPUT 32 "RegMDROut";
    .port_info 2 /OUTPUT 32 "LSControlOut";
v00000286ea6e8760_0 .net "LSControl", 1 0, v00000286ea6e6c00_0;  1 drivers
v00000286ea6e8800_0 .var "LSControlOut", 31 0;
v00000286ea6e9570_0 .net "RegMDROut", 31 0, v00000286ea6ec640_0;  1 drivers
E_00000286ea6e4960 .event anyedge, v00000286ea6e8760_0, v00000286ea6e9570_0;
S_00000286ea6e7c10 .scope module, "uut_ss" "ss" 2 28, 4 15 0, S_00000286ea6ef4b0;
 .timescale -9 -12;
    .port_info 0 /INPUT 2 "RegSSControl";
    .port_info 1 /INPUT 32 "RegBOut";
    .port_info 2 /INPUT 32 "RegMDROut";
    .port_info 3 /OUTPUT 32 "SSControlOut";
v00000286ea6e9610_0 .net "RegBOut", 31 0, v00000286ea6ec5a0_0;  1 drivers
v00000286ea6ed5a0_0 .net "RegMDROut", 31 0, v00000286ea6ec6e0_0;  1 drivers
v00000286ea6ed640_0 .net "RegSSControl", 1 0, v00000286ea6ec780_0;  1 drivers
v00000286ea6e6b60_0 .var "SSControlOut", 31 0;
E_00000286ea6f4cc0 .event anyedge, v00000286ea6ed640_0, v00000286ea6ed5a0_0, v00000286ea6e9610_0;
    .scope S_00000286ea6abda0;
T_0 ;
    %wait E_00000286ea6e4960;
    %load/vec4 v00000286ea6e8760_0;
    %dup/vec4;
    %pushi/vec4 0, 0, 2;
    %cmp/u;
    %jmp/1 T_0.0, 6;
    %dup/vec4;
    %pushi/vec4 1, 0, 2;
    %cmp/u;
    %jmp/1 T_0.1, 6;
    %dup/vec4;
    %pushi/vec4 2, 0, 2;
    %cmp/u;
    %jmp/1 T_0.2, 6;
    %pushi/vec4 0, 0, 32;
    %store/vec4 v00000286ea6e8800_0, 0, 32;
    %jmp T_0.4;
T_0.0 ;
    %pushi/vec4 0, 0, 24;
    %load/vec4 v00000286ea6e9570_0;
    %parti/s 8, 0, 2;
    %concat/vec4; draw_concat_vec4
    %store/vec4 v00000286ea6e8800_0, 0, 32;
    %jmp T_0.4;
T_0.1 ;
    %pushi/vec4 0, 0, 16;
    %load/vec4 v00000286ea6e9570_0;
    %parti/s 16, 0, 2;
    %concat/vec4; draw_concat_vec4
    %store/vec4 v00000286ea6e8800_0, 0, 32;
    %jmp T_0.4;
T_0.2 ;
    %load/vec4 v00000286ea6e9570_0;
    %store/vec4 v00000286ea6e8800_0, 0, 32;
    %jmp T_0.4;
T_0.4 ;
    %pop/vec4 1;
    %jmp T_0;
    .thread T_0, $push;
    .scope S_00000286ea6e7c10;
T_1 ;
    %wait E_00000286ea6f4cc0;
    %load/vec4 v00000286ea6ed640_0;
    %dup/vec4;
    %pushi/vec4 0, 0, 2;
    %cmp/u;
    %jmp/1 T_1.0, 6;
    %dup/vec4;
    %pushi/vec4 1, 0, 2;
    %cmp/u;
    %jmp/1 T_1.1, 6;
    %dup/vec4;
    %pushi/vec4 2, 0, 2;
    %cmp/u;
    %jmp/1 T_1.2, 6;
    %load/vec4 v00000286ea6ed5a0_0;
    %store/vec4 v00000286ea6e6b60_0, 0, 32;
    %jmp T_1.4;
T_1.0 ;
    %load/vec4 v00000286ea6ed5a0_0;
    %parti/s 24, 8, 5;
    %load/vec4 v00000286ea6e9610_0;
    %parti/s 8, 0, 2;
    %concat/vec4; draw_concat_vec4
    %store/vec4 v00000286ea6e6b60_0, 0, 32;
    %jmp T_1.4;
T_1.1 ;
    %load/vec4 v00000286ea6ed5a0_0;
    %parti/s 16, 16, 6;
    %load/vec4 v00000286ea6e9610_0;
    %parti/s 16, 0, 2;
    %concat/vec4; draw_concat_vec4
    %store/vec4 v00000286ea6e6b60_0, 0, 32;
    %jmp T_1.4;
T_1.2 ;
    %load/vec4 v00000286ea6e9610_0;
    %store/vec4 v00000286ea6e6b60_0, 0, 32;
    %jmp T_1.4;
T_1.4 ;
    %pop/vec4 1;
    %jmp T_1;
    .thread T_1, $push;
    .scope S_00000286ea6ef4b0;
T_2 ;
    %vpi_call 2 37 "$dumpfile", "test_ls_ss_units.vcd" {0 0 0};
    %vpi_call 2 38 "$dumpvars", 32'sb00000000000000000000000000000000, S_00000286ea6ef4b0 {0 0 0};
    %end;
    .thread T_2;
    .scope S_00000286ea6ef4b0;
T_3 ;
    %vpi_call 2 43 "$display", "=== TESTE DAS UNIDADES LS E SS ADAPTADAS ===" {0 0 0};
    %vpi_call 2 46 "$display", "\012=== TESTE DA UNIDADE LS (Load Size) ===" {0 0 0};
    %pushi/vec4 2864434397, 0, 32;
    %store/vec4 v00000286ea6ec640_0, 0, 32;
    %pushi/vec4 0, 0, 2;
    %store/vec4 v00000286ea6e6c00_0, 0, 2;
    %delay 10000, 0;
    %vpi_call 2 54 "$display", "Load Byte (LSControl=00):" {0 0 0};
    %vpi_call 2 55 "$display", "  RegMDROut: 0x%h", v00000286ea6ec640_0 {0 0 0};
    %vpi_call 2 56 "$display", "  LSControlOut: 0x%h (esperado: 0x000000DD)", v00000286ea6e6ca0_0 {0 0 0};
    %pushi/vec4 1, 0, 2;
    %store/vec4 v00000286ea6e6c00_0, 0, 2;
    %delay 10000, 0;
    %vpi_call 2 61 "$display", "Load Halfword (LSControl=01):" {0 0 0};
    %vpi_call 2 62 "$display", "  RegMDROut: 0x%h", v00000286ea6ec640_0 {0 0 0};
    %vpi_call 2 63 "$display", "  LSControlOut: 0x%h (esperado: 0x0000CCDD)", v00000286ea6e6ca0_0 {0 0 0};
    %pushi/vec4 2, 0, 2;
    %store/vec4 v00000286ea6e6c00_0, 0, 2;
    %delay 10000, 0;
    %vpi_call 2 68 "$display", "Load Word (LSControl=10):" {0 0 0};
    %vpi_call 2 69 "$display", "  RegMDROut: 0x%h", v00000286ea6ec640_0 {0 0 0};
    %vpi_call 2 70 "$display", "  LSControlOut: 0x%h (esperado: 0xAABBCCDD)", v00000286ea6e6ca0_0 {0 0 0};
    %vpi_call 2 73 "$display", "\012=== TESTE DA UNIDADE SS (Store Size) ===" {0 0 0};
    %pushi/vec4 305419896, 0, 32;
    %store/vec4 v00000286ea6ec5a0_0, 0, 32;
    %pushi/vec4 2864434397, 0, 32;
    %store/vec4 v00000286ea6ec6e0_0, 0, 32;
    %pushi/vec4 0, 0, 2;
    %store/vec4 v00000286ea6ec780_0, 0, 2;
    %delay 10000, 0;
    %vpi_call 2 82 "$display", "Store Byte (SSControl=00):" {0 0 0};
    %vpi_call 2 83 "$display", "  RegBOut: 0x%h", v00000286ea6ec5a0_0 {0 0 0};
    %vpi_call 2 84 "$display", "  RegMDROut: 0x%h", v00000286ea6ec6e0_0 {0 0 0};
    %vpi_call 2 85 "$display", "  SSControlOut: 0x%h (esperado: 0xAABBCC78)", v00000286ea6ec820_0 {0 0 0};
    %pushi/vec4 1, 0, 2;
    %store/vec4 v00000286ea6ec780_0, 0, 2;
    %delay 10000, 0;
    %vpi_call 2 90 "$display", "Store Halfword (SSControl=01):" {0 0 0};
    %vpi_call 2 91 "$display", "  RegBOut: 0x%h", v00000286ea6ec5a0_0 {0 0 0};
    %vpi_call 2 92 "$display", "  RegMDROut: 0x%h", v00000286ea6ec6e0_0 {0 0 0};
    %vpi_call 2 93 "$display", "  SSControlOut: 0x%h (esperado: 0xAABB5678)", v00000286ea6ec820_0 {0 0 0};
    %pushi/vec4 2, 0, 2;
    %store/vec4 v00000286ea6ec780_0, 0, 2;
    %delay 10000, 0;
    %vpi_call 2 98 "$display", "Store Word (SSControl=10):" {0 0 0};
    %vpi_call 2 99 "$display", "  RegBOut: 0x%h", v00000286ea6ec5a0_0 {0 0 0};
    %vpi_call 2 100 "$display", "  RegMDROut: 0x%h", v00000286ea6ec6e0_0 {0 0 0};
    %vpi_call 2 101 "$display", "  SSControlOut: 0x%h (esperado: 0x12345678)", v00000286ea6ec820_0 {0 0 0};
    %vpi_call 2 104 "$display", "\012=== TESTE DE CASOS EXTREMOS ===" {0 0 0};
    %pushi/vec4 0, 0, 32;
    %store/vec4 v00000286ea6ec640_0, 0, 32;
    %pushi/vec4 0, 0, 32;
    %store/vec4 v00000286ea6ec5a0_0, 0, 32;
    %pushi/vec4 4294967295, 0, 32;
    %store/vec4 v00000286ea6ec6e0_0, 0, 32;
    %pushi/vec4 0, 0, 2;
    %store/vec4 v00000286ea6e6c00_0, 0, 2;
    %pushi/vec4 0, 0, 2;
    %store/vec4 v00000286ea6ec780_0, 0, 2;
    %delay 10000, 0;
    %vpi_call 2 114 "$display", "Load Byte de zero: 0x%h (esperado: 0x00000000)", v00000286ea6e6ca0_0 {0 0 0};
    %vpi_call 2 115 "$display", "Store Byte de zero: 0x%h (esperado: 0xFFFFFF00)", v00000286ea6ec820_0 {0 0 0};
    %pushi/vec4 4294967295, 0, 32;
    %store/vec4 v00000286ea6ec640_0, 0, 32;
    %pushi/vec4 4294967295, 0, 32;
    %store/vec4 v00000286ea6ec5a0_0, 0, 32;
    %pushi/vec4 0, 0, 32;
    %store/vec4 v00000286ea6ec6e0_0, 0, 32;
    %pushi/vec4 2, 0, 2;
    %store/vec4 v00000286ea6e6c00_0, 0, 2;
    %pushi/vec4 2, 0, 2;
    %store/vec4 v00000286ea6ec780_0, 0, 2;
    %delay 10000, 0;
    %vpi_call 2 125 "$display", "Load Word m\303\241ximo: 0x%h (esperado: 0xFFFFFFFF)", v00000286ea6e6ca0_0 {0 0 0};
    %vpi_call 2 126 "$display", "Store Word m\303\241ximo: 0x%h (esperado: 0xFFFFFFFF)", v00000286ea6ec820_0 {0 0 0};
    %vpi_call 2 128 "$display", "\012=== TESTE DAS UNIDADES LS E SS CONCLU\303\215DO ===" {0 0 0};
    %vpi_call 2 129 "$finish" {0 0 0};
    %end;
    .thread T_3;
# The file index is used to find the file name in the following table.
:file_names 5;
    "N/A";
    "<interactive>";
    "test_ls_ss_units.v";
    "../ls.v";
    "../ss.v";
